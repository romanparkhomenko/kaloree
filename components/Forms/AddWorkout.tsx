import React, { useState } from 'react';

import { Card, CardBody, CardTitle, Container, Row, Col, Form, FormGroup, Input } from 'reactstrap';
import Router, { useRouter } from 'next/router';

const AddWorkout: React.FC = () => {
  const [workout, setWorkout] = useState('');
  const [minutes, setMinutes] = useState('');
  const [caloriesBurnt, setCaloriesBurnt] = useState('');
  const [weight, setWeight] = useState('');
  const [successMessage, setSuccessMessage] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  const submitData = async (e: React.SyntheticEvent) => {
    e.preventDefault();
    setSuccessMessage('');
    setErrorMessage('');
    try {
      const body = { workout, minutes, caloriesBurnt, weight };
      await fetch(`/api/workout`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      })
        .then(res => res.json())
        .then(response => {
          console.info(response);
          setSuccessMessage(`Successfully logged ${workout}`);
          setWorkout('');
          setMinutes('');
          setCaloriesBurnt('');
          setWeight('');
          setTimeout(() => {
            setSuccessMessage('');
          }, 5000);
        })
        .catch(error => {
          setErrorMessage('Whoops, something went wrong');
          console.info(error);
        });
    } catch (error) {
      setErrorMessage('Whoops, something went wrong');
      console.error(error);
    }
  };

  return (
    <>
      <Card className="card-stats mb-4 mb-xl-0">
        <CardBody>
          <Row>
            <div className="col">
              <Form>
                <Col lg="12" className="align-items-center justify-content-between row">
                  <h6 className="heading-small text-muted mb-4">Log a Workout</h6>
                  <span className="icon icon-shape bg-gradient-orange text-white rounded-circle shadow">
                    <i className="ni ni-user-run" />
                  </span>
                </Col>
                <Row>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="workout">
                      Workout
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="workout"
                      placeholder="What kind of workout did you do?"
                      onChange={e => setWorkout(e.target.value)}
                      value={workout}
                      type="text"
                    />
                  </Col>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="minutes">
                      Minutes
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="minutes"
                      placeholder="How long did you workout?"
                      onChange={e => setMinutes(e.target.value)}
                      value={minutes}
                      type="number"
                    />
                  </Col>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="caloriesBurnt">
                      Calories Burnt
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="caloriesBurnt"
                      placeholder="Calories Burnt"
                      onChange={e => setCaloriesBurnt(e.target.value)}
                      value={caloriesBurnt}
                      type="number"
                    />
                  </Col>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="weight">
                      Weight
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="weight"
                      placeholder="How much weight did you lift?"
                      onChange={e => setWeight(e.target.value)}
                      value={weight}
                      type="number"
                    />
                  </Col>
                </Row>
                <Row className="align-items-center justify-content-end">
                  <Col lg="12" className="mt-3">
                    <button
                      className="btn btn-info bg-gradient-orange w-100"
                      onClick={e => submitData(e)}
                    >
                      Log My Workout
                    </button>
                    {successMessage && (
                      <p className="text-center font-weight-700 text-info">{successMessage}</p>
                    )}
                    {errorMessage && (
                      <p className="text-center font-weight-700 text-warning">{errorMessage}</p>
                    )}
                  </Col>
                </Row>
              </Form>
            </div>
          </Row>
        </CardBody>
      </Card>
    </>
  );
};

export default AddWorkout;
