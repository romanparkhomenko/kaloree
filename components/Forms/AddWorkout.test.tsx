import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AddWorkout from './AddWorkout';

describe('Add Workout Form', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<AddWorkout />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('All inputs work and user can submit data', async () => {
    render(<AddWorkout />);

    const workoutInput = screen.getByPlaceholderText('What kind of workout did you do?');
    const minutesInput = screen.getByPlaceholderText('How long did you workout?');
    const caloriesBurntInput = screen.getByPlaceholderText('Calories Burnt');
    const weightInput = screen.getByPlaceholderText('How much weight did you lift?');
    const submitButton = screen.getByText('Log My Workout');

    fireEvent.change(workoutInput, { target: { value: 'Deadlift' } });
    expect(workoutInput).toHaveValue('Deadlift');

    fireEvent.change(minutesInput, { target: { value: '60' } });
    expect(minutesInput).toHaveValue(60);

    fireEvent.change(caloriesBurntInput, { target: { value: '400' } });
    expect(caloriesBurntInput).toHaveValue(400);

    fireEvent.change(weightInput, { target: { value: '200' } });
    expect(weightInput).toHaveValue(200);

    const fakeWeightResponse = { okay: 'workout logged' };

    global.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeWeightResponse),
      });
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    // Confirm success message exists and values are reset
    expect(screen.getByText('Successfully logged Deadlift')).toBeVisible();
    expect(workoutInput).toHaveValue('');
    expect(minutesInput).toHaveValue(null);
    expect(caloriesBurntInput).toHaveValue(null);
    expect(weightInput).toHaveValue(null);
  });
});
