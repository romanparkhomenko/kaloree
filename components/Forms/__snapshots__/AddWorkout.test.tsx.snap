// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Workout Form renders correctly 1`] = `
<div
  className="card-stats mb-4 mb-xl-0 card"
>
  <div
    className="card-body"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <form
          className=""
        >
          <div
            className="align-items-center justify-content-between row col-lg-12"
          >
            <h6
              className="heading-small text-muted mb-4"
            >
              Log a Workout
            </h6>
            <span
              className="icon icon-shape bg-gradient-orange text-white rounded-circle shadow"
            >
              <i
                className="ni ni-user-run"
              />
            </span>
          </div>
          <div
            className="row"
          >
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="workout"
              >
                Workout
              </label>
              <input
                className="form-control-alternative form-control"
                id="workout"
                onChange={[Function]}
                placeholder="What kind of workout did you do?"
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="minutes"
              >
                Minutes
              </label>
              <input
                className="form-control-alternative form-control"
                id="minutes"
                onChange={[Function]}
                placeholder="How long did you workout?"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="caloriesBurnt"
              >
                Calories Burnt
              </label>
              <input
                className="form-control-alternative form-control"
                id="caloriesBurnt"
                onChange={[Function]}
                placeholder="Calories Burnt"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="weight"
              >
                Weight
              </label>
              <input
                className="form-control-alternative form-control"
                id="weight"
                onChange={[Function]}
                placeholder="How much weight did you lift?"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="align-items-center justify-content-end row"
          >
            <div
              className="mt-3 col-lg-12"
            >
              <button
                className="btn btn-info bg-gradient-orange w-100"
                onClick={[Function]}
              >
                Log My Workout
              </button>
              
              
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
